internal/migration/migration.go

```diff
@@ -0,0 +1,406 @@
+	logger.Warnw("buffer size set to 0. Running automatic buffer inference. Make sure you have enough system memory for this")
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return -1, err
+	}
+	defer existingDB.Close()
+
+	var bufferSize = 0
+
+	err = existingDB.View(func(tx *bbolt.Tx) error {
+		bufferSize = tx.Bucket(bucketName).Stats().KeyN
+		return nil
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	logger.Infow("buffer size automatically inferred from existing DB", "bufferSize", bufferSize)
+
+	return bufferSize, nil
+}
+
+func shouldMigrate(logger log.Logger, sourceBeaconPath, beaconName string, destination chain.StorageType, pgDSN string) error {
+	//nolint:exhaustive // We want to explicitly ignore the chain.MemDB backend since there's nothing to migrate there.
+	switch destination {
+	case chain.BoltDB:
+		return shouldMigrateBolt(sourceBeaconPath)
+	case chain.PostgreSQL:
+		return shouldMigratePostgres(logger, beaconName, pgDSN)
+	default:
+		return fmt.Errorf("unknown destination type %q for migration package", destination)
+	}
+}
+
+func shouldMigrateBolt(sourceBeaconPath string) error {
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return err
+	}
+	defer existingDB.Close()
+
+	return existingDB.View(func(tx *bbolt.Tx) error {
+		_, value := tx.Bucket(bucketName).Cursor().First()
+		var b = beacon{}
+		if json.Unmarshal(value, &b) != nil {
+			return ErrMigrationNotNeeded
+		}
+
+		return nil
+	})
+}
+
+func pgConn(ctx context.Context, logger log.Logger, beaconName, pgDSN string) (*pgdb.Store, func(), error) {
+	pgConf, err := database.ConfigFromDSN(pgDSN)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	conn, err := database.Open(ctx, pgConf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := schema.Migrate(ctx, conn); err != nil {
+		err := fmt.Errorf("migrating error: %w", err)
+		return nil, nil, err
+	}
+
+	store, err := pgdb.NewStore(ctx, logger, conn, beaconName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cancel := func() {
+		store.Close(ctx)
+		conn.Close()
+	}
+
+	return store, cancel, err
+}
+
+func shouldMigratePostgres(logger log.Logger, beaconName, pgDSN string) error {
+	ctx := context.Background()
+
+	store, cancel, err := pgConn(ctx, logger, beaconName, pgDSN)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	storeLen, err := store.Len(ctx)
+	if err != nil {
+		return err
+	}
+
+	if storeLen != 0 {
+		logger.Warnw("Postgres contains beacon rounds. Skipping migration", "rounds count", storeLen)
+		return ErrMigrationNotNeeded
+	}
+
+	return nil
+}
+
+func (m *migrator) doMigrate() (retErr error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer func() {
+		cancel()
+
+		finishedIn := time.Since(m.startedAt).String()
+
+		if m.existingRows != m.migratedRows {
+			retErr = fmt.Errorf("not all rounds migrated successfully expected: %d actual: %d", m.existingRows, m.migratedRows)
```

Comment:
We might want to add something like:
```suggestion
			retErr = fmt.Errorf("not all rounds migrated successfully expected: %d actual: %d\nconsider running drand sync to restore the missing rounds", m.existingRows, m.migratedRows)
```